### PR TITLE
WIP: Aggregate RCOutput with UAVCAN to one place in SRV_Channels

### PR DESCRIPTION
**Pull request by EShamaev**
As support for UAVCAN output is widening, there is a need to put it in one place instead of copy-paste same code all over libraries.
Maybe this is overkill, but solves the problem.
As a side effect adds UAVCAN support to Intel Aero as it came from https://github.com/ArduPilot/ardupilot/pull/7729

### CODE REVIEW

**Comment by anemostec** [comment]
Thanks Shamaev for this support. For the intel aero we need the RCoutput tap to be removed when the UAVCAN support is added. The reason for this is that RCouputtap uses the /dev/ttyS1 which could be used by other portions of the code, for the telemetry for instance.

**Comment by EShamaev** [comment]
It should work without a tap now, can you give it a try?

**Comment by anemostec** [comment]
not yet, I'm in the middle of an other development and CAN is not connected

**Comment by anemostec** [comment]
RCoutputTap should be desactivated, I know it can work without it

**Comment by EShamaev** [comment]
What I mean is that in this PR output is updated not in tap module but in normal RCOutput for Intel Aero. Tap is not needed and it's a change from your old PR. Please make tests when you can to make sure all is working.

**Comment by lucasdemarchi** [comment]
@EShamaev I gave it a quick look and it's much better now. However I think the individual RCOutput implementations should not be changed to call uavcan. Instead leave it to be done by whom call them. This way it's copter/plane/etc (or AP_Motor if everybody is using it) responsibility to do it and you then don't have to change each driver. 

Thanks.

**Comment by EShamaev** [comment]
Will it be enough to add it to SRV_Channels::push() then?

**Comment by EShamaev** [comment]
Actually UAVCAN can be totally separated from RCOutput and placed into SRV_Channels.

**Comment by lucasdemarchi** [comment]
Looks like a good option. I can't check right now as I'm on a phone. Just check if it cover all the callers.

File: libraries/AP_HAL_Linux/RCOutput_PCA9685.h [hunk 1]
```diff
@@ -10,6 +10,8 @@
 #define PCA9685_QUATENARY_ADDRESS           0x55
 #define PCA9685_QUINARY_ADDRESS             0x61
 
+#define PWM_CHAN_COUNT 16
```
Comment (khancyr) [inline libraries/AP_HAL_Linux/RCOutput_PCA9685.h]: this one should be put back into .cpp too

**Comment by khancyr** [comment]
Looks good, but I am still not sure where this must be called, on lib like SRV_Channels or on vehicle code directly ? I don't know ... I got the same problem with a serial protocol for rover motor ...

**Comment by EShamaev** [comment]
There is no such calls left in vehicles code. All is done through SRV_Channels.

**Comment by khancyr** [comment]
We can put that in the same place as SRV_Channels::output_ch_all() or do another lib that will handle the update of srv_chanell if enable, uavcan if enable, whatever_protocol if enable etc .
SRV_Channel is really for channel, I am not we want it to be polluted by can. But I don't have a better idea, so lets merge something and refactor atfer !

**Comment by EShamaev** [comment]
Is SRV_Channels a singleton?

**Comment by khancyr** [comment]
hummmm SRV_channel update sbus and voltz protocl so, let's put uavcan here too ^^

**Comment by EShamaev** [comment]
Yes, that's logical place for it to be.

File: libraries/SRV_Channel/SRV_Channel.h [hunk 2]
```diff
@@ -420,6 +420,7 @@ class SRV_Channels {
     static void cork();
 
     static void push();
+    static void push_UAVCAN(void);
```
Comment (magicrub) [inline libraries/SRV_Channel/SRV_Channel.h]: push_UAVCAN() should be private

**Comment by magicrub** [comment]
Looks good to me! Thanks @EShamaev !!!

File: libraries/SRV_Channel/SRV_Channel.h (hunk 2 above)
Comment (EShamaev) [reply libraries/SRV_Channel/SRV_Channel.h]: OK

**Comment by OXINARF** [comment]
Fantastic, less work for me! 🙂

I will argue though that this should be done the same as VOLZ and Sbus - meaning that calling an `update_servo_output` should be enough, with UAVCAN side retrieving the values. Only difference will be to calling `get_UAVCAN` on runtime in SRV_Channels, opposite to what is done with the other two.
If that's not done now, I'll have do it later because I'll be adding another CAN protocol that needs to send servo output out - and I don't want an unneeded mess of CAN code inside SRV_Channel.

**Comment by EShamaev** [comment]
SRV_Channels::push() is called from vehicle's code as the only instance of updating servos. So if we get away from RCOutput for UAVCAN, I can hardly imagine other place to put it. And I do not think that we need to call something from vehicle code that will take the data from SRV_Channels anyway.

**Comment by OXINARF** [comment]
@EShamaev You didn't understand my comment. Look at how AP_Volz and AP_SbusOut work; there are no calls from the vehicle code and it's a one line in SRV_Channels.

UAVCAN can't be one line, but can be much less code inside SRV_Channels (from line 237 to 254, that should all be hidden inside AP_UAVCAN).

**Comment by magicrub** [comment]
@OXINARF check it now, I moved a few things over to AP_UAVCAN

**Comment by lucasdemarchi** [comment]
@magicrub I think we can move more. I think @OXINARF was referring to something like https://github.com/lucasdemarchi/ardupilot/commit/31f694c0da756a4825f54a4dac611781cd51e70c - if so, I strongly agree with him :)

**Comment by magicrub** [comment]
Yup. Even better! If really like to remove those ifdef's everywhere. This empty class is a good way to do it. Thank you.

**Comment by OXINARF** [comment]
@lucasdemarchi @magicrub Actually I prefer what Tom did. The reason is simple: if that loop goes to the UAVCAN library then every CAN protocol will have to repeat it in its own library - iterating the CAN interfaces multiple times doesn't make sense to me.

**Comment by lucasdemarchi** [comment]
Do you plan to have different protocols on the same CAN interfaces?

**Comment by OXINARF** [comment]
What do you mean the same CAN interface? Each interface can run one protocol... I'm definitely going to do that.

**Comment by lucasdemarchi** [comment]
So each one should loop through only the interfaces it is managing. They are either not the same loop or CAN manager should be protocol-independent, not spread its loop over the entire codebase.

**Comment by OXINARF** [comment]
@lucasdemarchi if you have time to do those changes, be my guest, I don't have time to do it. AP_BoardConfig_CAN::get_can_num_ifaces will loop through the maximum number of interfaces to give the total number of interfaces available and that will be repeated in every library if we go with your option.

**Comment by lucasdemarchi** [comment]
Right now we have only uavcan, or am I missing something? When we have another one we can split the interfaces among them.

**Comment by OXINARF** [comment]
I've said multiple times that I'm developing code to add another CAN protocol... I won't have time to do split by interface, it's outside the scope of my work. If people want things done correctly, then they should be demanded when original code is done (instead of ignoring them so PRs don't bitrot), not ask them later to other devs.

**Comment by lucasdemarchi** [comment]
> I've said multiple times that I'm developing code to add another CAN protocol... I won't have time to do split by interface, it's outside the scope of my work. 

How can I do such a thing? When you post your code all you have to do is to add a loop for 2 interfaces in your library. Even if we have this in 2 libraries it's better than having it all spread in the rest.


> If people want things done correctly, then they should be demanded when original code is done (instead of ignoring them so PRs don't bitrot), not ask them later to other devs

We disagree with the solution and that's fine. But let's stay on topic instead of diverging.

**Comment by magicrub** [comment]
The AP_UAVCAN library allows additional protocols but only uavcan is implemented. There's a protocol param. So, AP_UAVCAN is poorly named

**Comment by OXINARF** [comment]
@magicrub I'm not sure what you are saying. AP_UAVCAN is for UAVCAN... Other protocols should have their own library.

**Comment by magicrub** [comment]
@OXINARF I meant that CAN_Dx_PROTOCOL will set the [CAN parsing algorithm](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp#L134) but there's always CAN_Dx_UC_ which[ always pulls from the AP_UAVCAN driver](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_BoardConfig/canbus_driver.cpp#L36) so even if you had _PROTOCOL=2 then you have to add it to AP_UAVCAN/AP_UAVCAN.cpp.

**Comment by OXINARF** [comment]
@magicrub No. CAN_Dx_UC_ will only be loaded for UAVCAN, not any other protocol. You don't have to add anything into AP_UAVCAN.cpp for other protocols.

**Comment by lucasdemarchi** [comment]
I do have boards with can, but I don't have any ESC right now with it. So, I think the only thing missing would be to test.

**Comment by tridge** [comment]
I have done a forced push to the PR to rebase it on master
I've also found it has a major memory leak - it drops to zero available memory as soon as it starts up on both HAL_PX4 and HAL_ChibiOS

**Pull request merged**